### ISSUE TITLE
As a User I want my data to be stored securely

### ISSUE THREAD

**Issue by noerw** [sitcomlab/simport-learning-app#9]
TBD what this means:
- data does not leave the device?
- data is encrypted?
  - the sqlite driver we use provides encryption!
- ...?

In general, what is the threat model?

**Comment by zven** [comment]
Background research: https://github.com/sitcomlab/simport-learning-app/issues/28

**Comment by felixerdy** [comment]
Currently implementing secure storage using [Capacitor sqlite](https://github.com/capacitor-community/sqlite) v2.9.x in [secure-storage branch](https://github.com/sitcomlab/simport-learning-app/tree/feature/secure-storage) 

TODO:
- [x] check migration from unencrypted db to encrypted db

**Comment by felixerdy** [comment]
Following error occurs on Android when updating from unsecure to secure db:
```
2021-03-02 15:26:33.747 10552-10634/de.ifgi.simport.learning E/Capacitor: Serious error executing plugin
    java.lang.reflect.InvocationTargetException
        at java.lang.reflect.Method.invoke(Native Method)
        at com.getcapacitor.PluginHandle.invoke(PluginHandle.java:99)
        at com.getcapacitor.Bridge$1.run(Bridge.java:521)
        at android.os.Handler.handleCallback(Handler.java:938)
        at android.os.Handler.dispatchMessage(Handler.java:99)
        at android.os.Looper.loop(Looper.java:223)
        at android.os.HandlerThread.run(HandlerThread.java:67)
     Caused by: net.sqlcipher.database.SQLiteException: file is not a database: , while compiling: select count(*) from sqlite_master;
        at net.sqlcipher.database.SQLiteCompiledSql.native_compile(Native Method)
        at net.sqlcipher.database.SQLiteCompiledSql.compile(SQLiteCompiledSql.java:89)
        at net.sqlcipher.database.SQLiteCompiledSql.<init>(SQLiteCompiledSql.java:62)
        at net.sqlcipher.database.SQLiteProgram.<init>(SQLiteProgram.java:91)
        at net.sqlcipher.database.SQLiteQuery.<init>(SQLiteQuery.java:48)
        at net.sqlcipher.database.SQLiteDirectCursorDriver.query(SQLiteDirectCursorDriver.java:60)
        at net.sqlcipher.database.SQLiteDatabase.rawQueryWithFactory(SQLiteDatabase.java:2016)
        at net.sqlcipher.database.SQLiteDatabase.rawQuery(SQLiteDatabase.java:1902)
        at net.sqlcipher.database.SQLiteDatabase.keyDatabase(SQLiteDatabase.java:2669)
        at net.sqlcipher.database.SQLiteDatabase.openDatabaseInternal(SQLiteDatabase.java:2599)
        at net.sqlcipher.database.SQLiteDatabase.openDatabase(SQLiteDatabase.java:1247)
        at net.sqlcipher.database.SQLiteDatabase.openDatabase(SQLiteDatabase.java:1214)
        at net.sqlcipher.database.SQLiteDatabase.openOrCreateDatabase(SQLiteDatabase.java:1304)
        at net.sqlcipher.database.SQLiteDatabase.openOrCreateDatabase(SQLiteDatabase.java:1292)
        at net.sqlcipher.database.SQLiteDatabase.openOrCreateDatabase(SQLiteDatabase.java:1284)
        at net.sqlcipher.database.SQLiteDatabase.openOrCreateDatabase(SQLiteDatabase.java:1329)
        at com.getcapacitor.community.database.sqlite.SQLite.Database.open(Database.java:132)
        at com.getcapacitor.community.database.sqlite.CapacitorSQLite.open(CapacitorSQLite.java:131)
        at java.lang.reflect.Method.invoke(Native Method) 
        at com.getcapacitor.PluginHandle.invoke(PluginHandle.java:99) 
        at com.getcapacitor.Bridge$1.run(Bridge.java:521) 
        at android.os.Handler.handleCallback(Handler.java:938) 
        at android.os.Handler.dispatchMessage(Handler.java:99) 
        at android.os.Looper.loop(Looper.java:223) 
        at android.os.HandlerThread.run(HandlerThread.java:67) 
2021-03-02 15:26:33.748 10552-10634/de.ifgi.simport.learning E/AndroidRuntime: FATAL EXCEPTION: CapacitorPlugins
    Process: de.ifgi.simport.learning, PID: 10552
    java.lang.RuntimeException: java.lang.reflect.InvocationTargetException
        at com.getcapacitor.Bridge$1.run(Bridge.java:530)
        at android.os.Handler.handleCallback(Handler.java:938)
        at android.os.Handler.dispatchMessage(Handler.java:99)
        at android.os.Looper.loop(Looper.java:223)
        at android.os.HandlerThread.run(HandlerThread.java:67)
     Caused by: java.lang.reflect.InvocationTargetException
        at java.lang.reflect.Method.invoke(Native Method)
        at com.getcapacitor.PluginHandle.invoke(PluginHandle.java:99)
        at com.getcapacitor.Bridge$1.run(Bridge.java:521)
        at android.os.Handler.handleCallback(Handler.java:938) 
        at android.os.Handler.dispatchMessage(Handler.java:99) 
        at android.os.Looper.loop(Looper.java:223) 
        at android.os.HandlerThread.run(HandlerThread.java:67) 
     Caused by: net.sqlcipher.database.SQLiteException: file is not a database: , while compiling: select count(*) from sqlite_master;
        at net.sqlcipher.database.SQLiteCompiledSql.native_compile(Native Method)
        at net.sqlcipher.database.SQLiteCompiledSql.compile(SQLiteCompiledSql.java:89)
        at net.sqlcipher.database.SQLiteCompiledSql.<init>(SQLiteCompiledSql.java:62)
        at net.sqlcipher.database.SQLiteProgram.<init>(SQLiteProgram.java:91)
        at net.sqlcipher.database.SQLiteQuery.<init>(SQLiteQuery.java:48)
        at net.sqlcipher.database.SQLiteDirectCursorDriver.query(SQLiteDirectCursorDriver.java:60)
        at net.sqlcipher.database.SQLiteDatabase.rawQueryWithFactory(SQLiteDatabase.java:2016)
        at net.sqlcipher.database.SQLiteDatabase.rawQuery(SQLiteDatabase.java:1902)
        at net.sqlcipher.database.SQLiteDatabase.keyDatabase(SQLiteDatabase.java:2669)
        at net.sqlcipher.database.SQLiteDatabase.openDatabaseInternal(SQLiteDatabase.java:2599)
        at net.sqlcipher.database.SQLiteDatabase.openDatabase(SQLiteDatabase.java:1247)
        at net.sqlcipher.database.SQLiteDatabase.openDatabase(SQLiteDatabase.java:1214)
        at net.sqlcipher.database.SQLiteDatabase.openOrCreateDatabase(SQLiteDatabase.java:1304)
        at net.sqlcipher.database.SQLiteDatabase.openOrCreateDatabase(SQLiteDatabase.java:1292)
        at net.sqlcipher.database.SQLiteDatabase.openOrCreateDatabase(SQLiteDatabase.java:1284)
        at net.sqlcipher.database.SQLiteDatabase.openOrCreateDatabase(SQLiteDatabase.java:1329)
        at com.getcapacitor.community.database.sqlite.SQLite.Database.open(Database.java:132)
        at com.getcapacitor.community.database.sqlite.CapacitorSQLite.open(CapacitorSQLite.java:131)
        at java.lang.reflect.Method.invoke(Native Method) 
        at com.getcapacitor.PluginHandle.invoke(PluginHandle.java:99) 
        at com.getcapacitor.Bridge$1.run(Bridge.java:521) 
        at android.os.Handler.handleCallback(Handler.java:938) 
        at android.os.Handler.dispatchMessage(Handler.java:99) 
        at android.os.Looper.loop(Looper.java:223) 
        at android.os.HandlerThread.run(HandlerThread.java:67) 
2021-03-02 15:26:33.778 10552-10634/de.ifgi.simport.learning E/com.marianhello.logging.UncaughtExceptionLogger: FATAL EXCEPTION: CapacitorPluginsjava.lang.RuntimeException: java.lang.reflect.InvocationTargetException
        at com.getcapacitor.Bridge$1.run(Bridge.java:530)
        at android.os.Handler.handleCallback(Handler.java:938)
        at android.os.Handler.dispatchMessage(Handler.java:99)
        at android.os.Looper.loop(Looper.java:223)
        at android.os.HandlerThread.run(HandlerThread.java:67)
     Caused by: java.lang.reflect.InvocationTargetException: null
        at java.lang.reflect.Method.invoke(Native Method)
        at com.getcapacitor.PluginHandle.invoke(PluginHandle.java:99)
        at com.getcapacitor.Bridge$1.run(Bridge.java:521)
    	... 4 common frames omitted
    Caused by: net.sqlcipher.database.SQLiteException: file is not a database: , while compiling: select count(*) from sqlite_master;
        at net.sqlcipher.database.SQLiteCompiledSql.native_compile(Native Method)
        at net.sqlcipher.database.SQLiteCompiledSql.compile(SQLiteCompiledSql.java:89)
        at net.sqlcipher.database.SQLiteCompiledSql.<init>(SQLiteCompiledSql.java:62)
        at net.sqlcipher.database.SQLiteProgram.<init>(SQLiteProgram.java:91)
        at net.sqlcipher.database.SQLiteQuery.<init>(SQLiteQuery.java:48)
        at net.sqlcipher.database.SQLiteDirectCursorDriver.query(SQLiteDirectCursorDriver.java:60)
        at net.sqlcipher.database.SQLiteDatabase.rawQueryWithFactory(SQLiteDatabase.java:2016)
        at net.sqlcipher.database.SQLiteDatabase.rawQuery(SQLiteDatabase.java:1902)
        at net.sqlcipher.database.SQLiteDatabase.keyDatabase(SQLiteDatabase.java:2669)
        at net.sqlcipher.database.SQLiteDatabase.openDatabaseInternal(SQLiteDatabase.java:2599)
        at net.sqlcipher.database.SQLiteDatabase.openDatabase(SQLiteDatabase.java:1247)
        at net.sqlcipher.database.SQLiteDatabase.openDatabase(SQLiteDatabase.java:1214)
        at net.sqlcipher.database.SQLiteDatabase.openOrCreateDatabase(SQLiteDatabase.java:1304)
        at net.sqlcipher.database.SQLiteDatabase.openOrCreateDatabase(SQLiteDatabase.java:1292)
        at net.sqlcipher.database.SQLiteDatabase.openOrCreateDatabase(SQLiteDatabase.java:1284)
        at net.sqlcipher.database.SQLiteDatabase.openOrCreateDatabase(SQLiteDatabase.java:1329)
        at com.getcapacitor.community.database.sqlite.SQLite.Database.open(Database.java:132)
        at com.getcapacitor.community.database.sqlite.CapacitorSQLite.open(CapacitorSQLite.java:131)
    	... 7 common frames omitted
```

**Comment by felixerdy** [comment]
After further research, there might be an issue with the secret key:
- When using the sqlite library you have to [set the secret keys in the native code](https://github.com/capacitor-community/sqlite/blob/master/docs/API.md#defining-your-own-secret-and-newsecret-keys-encryption-only)
- Currently, [you (or the user) can not define the secrets at runtime](https://github.com/capacitor-community/sqlite/issues/88). I guess modifying the code would not be that hard

Setting a secret before building the app would be a security issue for me as all apps would use the same secret key. 

Do we want the users to set their secret key? Or do we want to generate one during runtime?

**Comment by felixerdy** [comment]
In general, applications are running in sandboxes on [iOS](https://support.apple.com/guide/security/security-of-runtime-process-sec15bfe098e/web) and [Android](https://source.android.com/security/app-sandbox). By default, apps can't interact with each other and have limited access to the OS. Therefore, an app can only access its own file directory (e.g. where the SQLite DB is located) without access for other installed apps.

On Android, the device has to be rooted or attackers need physical access (plus the passphrase for the lockscreen and enabled ADB) in order to access local files like the app file directory. On iOS, that seems to be even harder as [each file on the filesystem gets encrypted by default](https://support.apple.com/guide/security/data-protection-overview-secf6276da8a/1/web/1).

In conclusion, I guess we could just stay with the current solution as the app file directory should be safe as long as the user doesn't enable ADB etc.

Side note: I also took a look at the [Ionic Secure Storage (Enterprise Solution)](https://ionic.io/products/secure-storage) which also uses an [Encrypted SQLite Database](https://ionic.io/docs/secure-storage) in combination with the [Ionic Identity Vault (Enterprise Solition)](https://ionic.io/products/identity-vault) to [store sensitive information in secure locations on the device](https://ionic.io/docs/identity-vault). That's however a paid product which is not open-source, but still worth sharing.

**Comment by felixerdy** [comment]
wontfix for now
